Add metadata fields for mappings (content gap initiative) (#6933)
opensearch-project/documentation-website: _field-types/index.md

# Mappings and field types

Mappings tell OpenSearch how to store and index your documents and their fields. You can specify the data type for each field (for example, `year` as `date`) to make storage and querying more efficient. 

While dynamic mappings automatically add new data and fields, using explicit mappings is recommended. Explicit mappings let you define the exact structure and data types upfront. This helps maintain data consistency and optimize performance, especially for large datasets or high-volume indexing operations.

> Even if you are unsure about your data structure at first, switching to explicit mappings later is recommended once you understand your data better. This helps avoid potential performance problems and keeps your data consistent. For example, with explicit mappings, you can ensure that `year` is treated as text and `age` as an integer, instead of both being interpreted as integers by dynamic mapping.

Reviewer: This paragraph feels slightly redundant with "This helps maintain data consistency and optimize performance, especially for large datasets or high-volume indexing operations." from the previous paragraph. Maybe combine "For example, with explicit mappings, you can ensure that `year` is treated as text and `age` as an integer, instead of both being interpreted as integers by dynamic mapping," with the previous paragraph and delete the rest?

```suggestion
For example, with explicit mappings, you can ensure that `year` is treated as text and `age` as an integer, instead of both being interpreted as integers by dynamic mapping.
```

## Dynamic mapping

When you index a document, OpenSearch adds fields automatically with dynamic mapping. You can also explicitly add fields to an index mapping.

### Dynamic mapping types

Type | Description
:--- | :---
`null` | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if the field has no value.
`boolean` | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
`float` | A single-precision, 32-bit floating-point number.
`double` | A double-precision, 64-bit floating-point number.
`integer` | A signed 32-bit number.
`object` | Objects are standard JSON objects, which can have fields and mappings of their own. For example, a `movies` object can have additional properties such as `title`, `year`, and `director`.

> `array` | OpenSearch does not have a specific array data type. Arrays are represented as a set of values of the same data type (for example, integers or strings) associated with a field. When indexing, you can pass multiple values for a field, and OpenSearch will treat it as an array. Empty arrays are valid and recognized as array fields with zero elements---not as fields with no values. OpenSearch supports querying and filtering arrays, including checking for values, range queries, and array operations like concatenation and intersection. Nested arrays, containing complex objects or other arrays, are also supported for advanced data modeling. 

Author: ```suggestion
`array` | OpenSearch does not have a specific array data type. Arrays are represented as a set of values of the same data type (for example, integers or strings) associated with a field. When indexing, you can pass multiple values for a field, and OpenSearch will treat it as an array. Empty arrays are valid and recognized as array fields with zero elements---not as fields with no values. OpenSearch supports querying and filtering arrays, including checking for values, range queries, and array operations like concatenation and intersection. Nested arrays, which may contain complex objects or other arrays, can also be used for advanced data modeling. 
```

### Dynamic templates

Dynamic templates are used to define custom mappings for dynamically added fields based on data type, field name, or field path. They allow you to define a flexible schema for your data, which can automatically adapt to changes in the structure or format of the input data.

You can use the following syntax to define a dynamic mapping template:

```json
PUT index
{
  "mappings": {
    "dynamic_templates": [
        {
          "fields": {
            "mapping": {
              "type": "short"
            },
            "match_mapping_type": "string",
            "path_match": "status*"
          }
        }
    ]
  }
}
```
{% include copy-curl.html %}

This mapping configuration dynamically maps any field with a name starting with `status` (for example, `status_code`) to the `short` data type if the initial value provided during indexing is a string.

### Dynamic mapping parameters

> The `dynamic_templates` support the following parameters for matching conditions and mapping rules. The default vaule is `null`.

Reviewer: ```suggestion
The `dynamic_templates` support the following parameters for matching conditions and mapping rules. The default value is `null`.
```

If you're starting to build out your cluster and data, you may not know exactly how your data should be stored. In those cases, you can use dynamic mappings, which tell OpenSearch to dynamically add data and its fields. However, if you know exactly what types your data falls under and want to enforce that standard, then you can use explicit mappings.

For example, if you want to indicate that `year` should be of type `text` instead of an `integer`, and `age` should be an `integer`, you can do so with explicit mappings. By using dynamic mapping, OpenSearch might interpret both `year` and `age` as integers.

This documentation provides an example for how to create an index mapping and how to add a document to it that will get `ip_range` validated.

## Dynamic mapping

When you index a document, OpenSearch adds fields automatically with dynamic mapping. You can also explicitly add fields to an index mapping.

### Dynamic mapping types

Type | Description
:--- | :---
null | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.

> boolean | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`

Author: ```suggestion
Boolean | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
```

You can define how documents and their fields are stored and indexed by creating a _mapping_. The mapping specifies the list of fields for a document. Every field in the document has a _field type_, which defines the type of data the field contains. For example, you may want to specify that the `year` field should be of type `date`. To learn more, see [Supported field types]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/index/).

> If you're starting to build out your cluster and data, you may not know exactly how your data should be stored. In those cases, you can use dynamic mappings, which tell OpenSearch to dynamically add data and its fields. However, if you know exactly what types your data falls under and want to enforce that standard, then you can use explicit mappings.

Author: ```suggestion
If you're starting to build out your cluster and data, you may not know exactly how your data should be stored. In those cases, you can use dynamic mappings, which tell OpenSearch to dynamically add data and its fields. However, it's important to note that dynamic mappings can have performance implications, especially for large datasets or high-volume indexing operations.

While dynamic mappings provide convenience and flexibility, it's generally recommended to use explicit mappings whenever possible. With explicit mappings, you define the exact structure and data types of your fields upfront, enforcing a consistent standard across your data. This approach not only ensures data integrity but also optimizes performance by allowing OpenSearch to efficiently store and query your data.

If you know exactly what types your data falls under and want to enforce that standard, then you should use explicit mappings. Even if you're unsure about your data structure initially, it is advisable to switch to explicit mappings once you have a better understanding of your data to avoid potential performance issues and maintain data consistency.
```

Reviewer 2: LGTM

> If you're starting to build out your cluster and data, you may not know exactly how your data should be stored. In those cases, you can use dynamic mappings, which tell OpenSearch to dynamically add data and its fields. However, if you know exactly what types your data falls under and want to enforce that standard, then you can use explicit mappings.

Author: Revised

For example, if you want to indicate that `year` should be of type `text` instead of an `integer`, and `age` should be an `integer`, you can do so with explicit mappings. By using dynamic mapping, OpenSearch might interpret both `year` and `age` as integers.

This documentation provides an example for how to create an index mapping and how to add a document to it that will get `ip_range` validated.

## Dynamic mapping

When you index a document, OpenSearch adds fields automatically with dynamic mapping. You can also explicitly add fields to an index mapping.

### Dynamic mapping types

Type | Description
:--- | :---

> null | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.

Author: ```suggestion
`null` | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
```

> Boolean | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`

Author: ```suggestion
`boolean` | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
```

> `boolean` | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`

Author: ```suggestion
boolean | OpenSearch accepts `true` and `false` as Boolean values. An empty string is equal to `false.`
```

> `null` | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.

Author: ```suggestion
null | A `null` field can't be indexed or searched. When a field is set to null, OpenSearch behaves as if that field has no values.
```

float | A single-precision 32-bit floating point number.
double | A double-precision 64-bit floating point number.
integer | A signed 32-bit number.
object | Objects are standard JSON objects, which can have fields and mappings of their own. For example, a `movies` object can have additional properties such as `title`, `year`, and `director`.

> array | Arrays in OpenSearch can only store values of one type, such as an array of only integers or strings. Empty arrays are treated as though they are fields with no values.

Reviewer 2: We should highlight that there is no specific array type, and a set of values can be passed for fields

Author: Revised

> array | Arrays in OpenSearch can only store values of one type, such as an array of only integers or strings. Empty arrays are treated as though they are fields with no values.

Author: ```suggestion
array | OpenSearch does not have a specific array data type. Arrays are represented as a set of values of the same data type (for example, integers or strings) associated with a field. When indexing, you can pass multiple values for a field, and OpenSearch will treat it as an array. Empty arrays are valid and recognized as array fields with zero elements, not as fields with no values. OpenSearch supports querying and filtering arrays, including checking for values, range queries, and array operations like concatenation and intersection. Nested arrays, containing complex objects or other arrays, are also supported for advanced data modeling. 
```

text | A string sequence of characters that represent full-text values.
keyword | A string sequence of structured characters, such as an email address or ZIP code.
date detection string | Enabled by default, if new string fields match a date's format, then the string is processed as a `date` field. For example, `date: "2012/03/11"` is processed as a date.
numeric detection string | If disabled, OpenSearch may automatically process numeric values as strings when they should be processed as numbers. When enabled, OpenSearch can process strings into `long`, `integer`, `short`, `byte`, `double`, `float`, `half_float`, `scaled_float`, and `unsigned_long`. Default is disabled.

> 

Reviewer 2: Can we cover all field types from here? https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesModule.java#L151-L174

Author: Should we link the reader to the GitHub file instead of listing the mapping types in the documentation?

Author: Should we link to the GitHub link instead of listing the field types in the documentation?

Reviewer 2: Linking to code can be tricky as the code can move due to refactoring making the documentation point to old/unusable links. Hence, I'd prefer to keep it within the documentation if possible.

### Dynamic templates

Dynamic templates are used to define custom mappings for dynamically added fields based on data type, field name, or field path. They allow you to define a flexible schema for your data, which can automatically adapt to changes in the structure or format of the input data.

The syntax for defining a dynamic mapping template in OpenSearch looks like the following:

```json
  "dynamic_templates": [
    {
      "template_name": { 
        ...  match conditions ... 
        "mapping": { ... } 
      }
    },
    ...
  ]

> ```

Reviewer 2: I can share an actual example here. Let me know if that will be better

Author: Yes, please share an example.

Reviewer 2: ```
PUT index
{
  "mappings": {
    "dynamic_templates": [
        {
          "fields": {
            "mapping": {
              "type": "short"
            },
            "match_mapping_type": "string",
            "path_match": "status*"
          }
        }
    ]
  }
}
```

The above will create a field like `status_code` (based on path_match parameter) with data type `short` if during the first pass, it detects the input value to be a `string`

{% include copy-curl.html %}

Note the following: 

- The template name can be any string value. 

> - The match conditions can include any of the following: `match_mapping_type`, `match`, `match_pattern`, `unmatch`, `path_match`, or `path_unmatch`. 

Reviewer 2: The supported params for matching conditions with `dynamic_templates` are:

`match_mapping_type`: Detected JSON type based on which mapping should be applied. (e.g. string, long, double, object, binary, boolean, date)
`match`: Regex for matching field name and apply mapping
`unmatch`: Regex to verify and not apply mapping if the field name satisfies the provided regex
`match_pattern`: Allowed values are `regex` and `simple`. Based on this, it applies the regex matching. Default is `simple`
`path_match`: Allows for nested path matching along with regex
`path_unmatch`: Allows for nested path matching along with regex
`mapping`: Mapping to be applied

Author: Added.

numeric detection string | If disabled, OpenSearch may automatically process numeric values as strings when they should be processed as numbers. When enabled, OpenSearch can process strings into `long`, `integer`, `short`, `byte`, `double`, `float`, `half_float`, `scaled_float`, and `unsigned_long`. Default is disabled.

### Dynamic templates

Dynamic templates are used to define custom mappings for dynamically added fields based on data type, field name, or field path. They allow you to define a flexible schema for your data, which can automatically adapt to changes in the structure or format of the input data.

The syntax for defining a dynamic mapping template in OpenSearch looks like the following:

```json
  "dynamic_templates": [
    {
      "template_name": { 
        ...  match conditions ... 
        "mapping": { ... } 
      }
    },
    ...
  ]
```
{% include copy-curl.html %}

Note the following: 

- The template name can be any string value. 
- The match conditions can include any of the following: `match_mapping_type`, `match`, `match_pattern`, `unmatch`, `path_match`, or `path_unmatch`. 
- Specify the `mapping` to be used for the matched field.

> 

Author: @mgodwan What are the defaults for these parameters?

Reviewer 2: Defaults are null. These are mostly logical wrapper which provide some kind of filters (based on the description). If they are not provided, the filter is not used for determining if dynamic template should be used for a field or not.

You cannot change the mapping of an existing field, you can only modify the field's mapping parameters. 
{: .note}

## Mapping parameters

Mapping parameters are used to configure the behavior of fields in an index. The following table lists commonly used mapping parameters.

Parameter | Description
:--- | :---
`analyzer` | Specifies the analyzer used to analyze string fields.
`boost` | Specifies a field-level query time to boost.
`coerce` | Tries to convert the value to the specified data type.
`copy_to` | Copies the values of this field to another field.
`doc_values` | Specifies whether the field should be stored on disk to make sorting and aggregation faster.
`dynamic` | Determines whether new fields should be added dynamically.
`enabled` | Specifies whether the field is enabled or disabled. 
`format` | Specifies the date format for date fields. 
`ignore_above` | Skips indexing values that are longer than the specified length.
`ignore_malformed` | Specifies whether malformed values should be ignored. 
`index` | Specifies whether the field should be indexed.

> `index_options` | Specifies what information should be stored in the index for scoring purposes.

Reviewer 2: do you think it will be helpful to add possible values for each of these parameters (or examples in case the value set is not limited)

Author: @mgodwan Yes to both.

Author: What are the default values?

Reviewer 2: The default values for each of these may vary across different field types (e.g. keyword, text may have different defaults for different parameters)

For each field type documentation (e.g. text, integer, geopoint), we can then share what the default/allowed values for the parameters are.

Author: @mgodwan Please verify the default and allowed values are accurate and make any necessary changes, as needed. Thank you.

## Mapping limit settings

OpenSearch may have certain limits or settings related to mappings, such as the maximum number of fields allowed in an index or the maximum depth of nested objects. These settings can be configured based on your requirements. <_Need SME input about mapping limits settings, their default values, and how to configure them if needed._>

## Removal of mapping types

<_Need SME input. Does OpenSearch allow for defining mapping types within an index? If so, SME needs to provide information to include in documentation. Do we have guidance on how to structure data without relying on mapping types?_>

## Runtime fields

> You can define fields at query time, rather than at index time, by using runtime fields. this can be useful for creating fields based on the values of other fields, or for performing transformations on data during the query process. Runtime fields are defined in the query itself and do not affect the underlying data in the index.

Reviewer 2: Ack. I can help you with the possible settings and examples (if applicable).

Reviewer 2: Mapping Limit Settings
 [TBU Description column]

| Settings | Default Value | Allowed Values | Type | Description |
|-|-|-|-|-|
| index.mapping.nested_fields.limit | 50 | [0,) | Dynamic | |
| index.mapping.nested_objects.limit | 10000 | [0,) | Dynamic | |
| index.mapping.total_fields.limit | 1000 | [0,) | Dynamic | |
| index.mapping.depth.limit | 20 | [1,100] | Dynamic | |
| index.mapping.field_name_length.limit | 50000 | [1,50000] | Dynamic | |
| index.mapper.dynamic | true | {true,false} | Dynamic | |

Author: @mgodwan  I added descriptions. Please confirm the descriptions are accurate.

| index.mapping.nested_objects.limit | 10000 | [0,) | Dynamic | Limits the maximum number of nested objects that can be created within a single document. |
| index.mapping.total_fields.limit | 1000 | [0,) | Dynamic | Limits the maximum number of fields that can be defined in an index mapping. |
| index.mapping.depth.limit | 20 | [1,100] | Dynamic | Limits the maximum depth of nested objects and nested fields that can be defined in an index mapping. |
| index.mapping.field_name_length.limit | 50000 | [1,50000] | Dynamic | Limits the maximum length of field names that can be defined in an index mapping. |
| index.mapper.dynamic | true | {true,false} | Dynamic | Determines whether new fields should be added dynamically to the mapping when they are encountered in a document. |

## Runtime fields

You can define fields at query time, rather than at index time, by using runtime fields. this can be useful for creating fields based on the values of other fields, or for performing transformations on data during the query process. Runtime fields are defined in the query itself and do not affect the underlying data in the index.

> 

Reviewer 2: This is not yet available on OpenSearch
https://github.com/opensearch-project/documentation-website/issues/6943

Author: Removed

  }
}
```
{% include copy-curl.html %}

You cannot change the mapping of an existing field, you can only modify the field's mapping parameters. 
{: .note}

## Mapping parameters

> Mapping parameters are used to configure the behavior of fields in an index. See the [Mapping parameters](inert-link-to-page) page for more information.

Author: ```suggestion
Mapping parameters are used to configure the behavior of fields in an index. See the [Mappings and field types]({{site.url}}{{site.baseurl}}//field-types/) page for more information.
```

> Mapping parameters are used to configure the behavior of fields in an index. See the [Mappings and field types]({{site.url}}{{site.baseurl}}//field-types/) page for more information.

Author: ```suggestion
Mapping parameters are used to configure the behavior of fields in an index. See the [Mappings and field types]({{site.url}}{{site.baseurl}}/field-types/) page for more information.
```

  }
}
```
{% include copy-curl.html %}

You cannot change the mapping of an existing field, you can only modify the field's mapping parameters. 
{: .note}

## Mapping parameters

Mapping parameters are used to configure the behavior of fields in an index. See [Mappings and field types]({{site.url}}{{site.baseurl}}/field-types/) for more information.

## Mapping limit settings

OpenSearch has certain limits or settings related to mappings, such as the settings listed in the following table. Settings can be configured based on your requirements. 

| Setting | Default value | Allowed value | Type | Description |
|-|-|-|-|-|

> | `index.mapping.nested_fields.limit` | 50 | [0,) | Dynamic | Limits the maximum number of nested fields that can be defined in an index mapping. |

Reviewer: Does `[0,)` mean that any value above `0` is allowed?

Author: @mgodwan Please see question above.

  }
}
```
{% include copy-curl.html %}

> You can add a document that has a malformed IP address to your index:

Author: Yes, we can move them.

```json
GET <index>/_mapping
```
{% include copy-curl.html %}

In the previous request, `<index>` may be an index name or a comma-separated list of index names. 

To get all mappings for all indexes, use the following request:

```json
GET _mapping
```
{% include copy-curl.html %}

To get a mapping for a specific field, provide the index name and the field name:

```json
GET _mapping/field/<fields>
GET /<index>/_mapping/field/<fields>
```
{% include copy-curl.html %}

> Both `<index>` and `<fields>` can be specified as one value or a comma-separated list. For example, the following request retrieves the mapping for the `year` and `age` fields in `sample-index1`:

Author: ```suggestion
Both `<index>` and `<fields>` can be specified as either one value or a comma-separated list. For example, the following request retrieves the mapping for the `year` and `age` fields in `sample-index1`:
```

  }
}
```
{% include copy-curl.html %}

## Mappings use cases

> See use case examples, including mapping string fields and ignoring malformed IP addresses, in [Mappings use cases](). 

Reviewer: ```suggestion
See use case examples, including mapping string fields and ignoring malformed IP addresses, in [Mappings use cases]({{site.url}}{{site.baseurl}}/field-types/mappings-use-cases/). 
```

  }
}
```

> ## Delete a mapping

Reviewer 2: I don't think we support `DELETE` verb on mappings.
